rust/server/src/handlers/create_user_handler.rs

```diff
@@ -0,0 +1,180 @@
+    let sessionless = Sessionless::new();
+
+    if let Ok(pub_key) = PublicKey::from_str(body.pub_key.as_str()) {
+        let sig = match Signature::from_str(body.signature.as_str()) {
+            Ok(s) => s,
+            Err(_) => {
+                return Json(Response::auth_error());
+            }
+        };
+
+        if sessionless.verify(message, &pub_key, &sig).is_err() {
+            return Json(Response::auth_error());
+        }
+
+        let key = PubKeys::key(&body.hash, &body.pub_key);
+        match data.user_client.clone().get_user_uuid(&key).await {
+            // If user exists with given (pub_key + hash), return back the user_uuid
+            Some(user_uuid) => Json(Response::user_success(user_uuid)),
+            None => {
+                // otherwise, put a new user
+                let new_uuid = Sessionless::generate_uuid();
+                match data.user_client.clone().put_user(&new_uuid.to_string(), &body.pub_key, &body.hash).await {
+                    Ok(user) => {
+                        // add pub key + hash with user uuid
+                        match data.user_client.clone().update_keys(&key, &user.uuid).await {
+                            Ok(_) => Json(Response::user_success(user.uuid)),
+                            Err(_) => Json(Response::server_error("Failed to update keys".to_string()))
+                        }
+                    },
+                    Err(_) => Json(Response::server_error("Failed to put user".to_string()))
+                }
+            }
+        }
+    } else {
+        return Json(Response::auth_error());
+    }
+}
+
+
+#[cfg(test)]
+mod tests {
+
+    use chrono::Utc;
+    use sessionless::Sessionless;
+
+    use crate::handlers::{CreateUserRequest, Response};
+    use crate::test_common::{self, check_path_exists, cleanup_test_files, setup_test_server, storage_uri};
+
+    #[tokio::test]
+    async fn test_create_user_handler() {
+        let stroage_uri = storage_uri("test_create_user_handler");
+        let test_server = setup_test_server(stroage_uri.clone());
+
+        assert!(test_server.is_running());
+        let sessionless = Sessionless::new();
+
+        let pub_key = sessionless.public_key();
+        let timestamp = Utc::now().timestamp().to_string();
+        let hash = "random_hash".to_string();
+
+        let message = format!("{}{}{}", timestamp, pub_key, hash);
+        let signature = sessionless.sign(message);
+
+        let payload = CreateUserRequest {
+            pub_key: pub_key.to_string(),
+            timestamp: timestamp,
+            hash: hash,
+            signature: signature.to_string(),
+        };
+
+
+        let response = test_server.post(test_common::USER_CREATE_PATH).json(&payload).await;
+
+        assert_eq!(response.clone().status_code(), 200);
+        // get the user_uuid from the response
+        // parse as Response
+        let user_resposne = response.json::<Response>();
```

Comment:
s/user_resposne/user_response/g ?

Reply:
Not sure what you meant but I noticed a typo. This is mainly test code though tbf